feat: added the ability to rotate guides in the canvas

mohamedAdhamc opened the pull request:
Fixes #2593 

This is the first step 
- now the rulers are able to rotate by holding control while moving the cursor (p.s vertical rulers have the center of rotation being half of their height, and horizontal have it as half their width which I think makes sense here if not though please let me know)


the main things left:
- handle moving the "rotated" rulers. 
- also if it would be required I could make a dialog which is accessed by right clicking  the guide to manually set the rotation.
- clean ups + optimizations

this is how the rotation works as of now:

https://user-images.githubusercontent.com/100296264/182651640-c854d38f-4893-4d37-8d19-9551989f5fa0.mp4

rodolforg commented:
It misses saving and loading the guide line angles (its position is stored in canvas metadata).

In a later PR, what about tooltips over guidelines? It could show:
- Current position (in synfig units and, if different, the unit set by user in preferences (pixel, mm, cm...)
- Current angle (in degrees)
- the tip of how to rotate it :) ("Press Ctrl and drag to rotate")

mohamedAdhamc commented:
> It misses saving and loading the guide line angles (its position is stored in canvas metadata).

ok I will look into storing the positions as canvas metadata when the remaining parts are finished, I did something similar for another feature (show/hide rulers) so I think the process should be kind of similar.


> In a later PR, what about tooltips over guidelines? It could show:
> 
> * Current position (in synfig units and, if different, the unit set by user in preferences (pixel, mm, cm...)
> * Current angle (in degrees)
> * the tip of how to rotate it :) ("Press Ctrl and drag to rotate")

this sounds like a good idea, but wouldn't it interfere (possible be a distraction) a bit with the canvas if there were many rulers ?

mohamedAdhamc commented:
ok finally I have handled detecting the current rotated ruler and moving the rotated ruler, there are just some small issues im handling and hopefully today I will be commiting. Also I'll try to save the meta data for the rotated rulers if I dont forget, it'll just involve saving the extra coordinates needed for the rotation. 

then after this what would be left is 
-the dialog on right click of rotated ruler
- optimizations/clean ups this PR is really messy as of right now xD
and as rodolf suggested then after this PR is good a tooltip can be implemented for it in another PR

mohamedAdhamc commented:
as explained in the commit description and previous comment, however, PR is really messy I will clean it up when I'm done.

mohamedAdhamc commented:
ok, so now the rotation and handling of the movement of the rotated ruler is almost done (needs more testing from my end and then feedback on the functionality from you guys but that would be after I finish it).

The main thing remaining now is the guide dialog which I laid the basis for in this last commit I will try to get it done and working tomorrow then maybe it'll take another day to clean up/test everything then it should be ready for review.

mohamedAdhamc commented:
ok, so its mostly done, mostly what remains is setting the guide through points in the dialog, setting by angle is done (although I think I'll add the option to change the degree tye from degree to rad). I'm heading out now then I'll be back so hopefully in a few hours before I sleep everything should be mostly done and maybe half the day tommrow will be spent optimizing/cleaning up the code is all over the place as of right now xD.

I kind of underestimated this one :). This so far has been the biggest feature in terms of size, it along with the very first are the largest so far, and the remaining shouldnt be this large so I think its fine to spend a little extra time reviewing it (tommrow) from my part before moving on to the remaingin three :).

mohamedAdhamc commented:
ok now I will be doing some cleaning up, and testing to get this finally ready for review.

However to summarize mostly what is done here is divided into couple main things:
1- rendering of the rotated ruler `renderer_guide.cpp`
2- handling storing the coordinates, and the events, flags, and rotation etc.. mostly `workarea`
3- then there  was finding the guide which was in `duckmatic`
5- also of course the guide dialog which was in `guide_dialog`
these are the main things that come to mind.

Regarding the implementation, the main idea was to set another point while rotating so that along with the center point we are able to draw the rotated line. and for moving the rotated ruler in very simple terms basically just adding an offset to the center and the other coordinate is determined by the mouse anyway.

mohamedAdhamc commented:
ok so just a reminder to myself here as I've been carried away a bit with previous features. what remains here is just finding a way to group the changing of the center and accompanying coordinates of the rotated ruler into one action using a passive group, along with a couple "comment removing" and this should be ready for review. I will do this first thing in the morning tomorrow after the "faster entry of values in value fields" feature.

mohamedAdhamc commented:
ok so I'm almost done making the passive group for the rotated ruler, i.e. making the undo and the history for the rotated ruler motion function properly. However, unfortunately the internet is down I'm using data xD. So I will finally mark this as ready for review and commit the remaining pat tomorrow morning.

rodolforg commented:
1. I don't have any idea what "accomp" abbreviation means :)
2. It breaks snap-to-guide
3- Anyway, I think you overly complicated things here. Why don't you use a new class (and metadata field) that stores, for each guideline, an anchor-point and an angle? Those two data already defines a unique straight line. `find_guide_x()` and `find_guide_y()` would be replaced with a single `find_guide()`

mohamedAdhamc commented:
> 1. I don't have any idea what "accomp" abbreviation means :)

I meant "accompanying", word is pretty long so I just abbreviated it like that not sure if its a common abbreviation xD.

> 2. It breaks snap-to-guide

oh yes thanks for pointing it out, actually I didn't know what snap to guide did, so I didnt handle snapping to guide when doing this. will do.

>    3- Anyway, I think you overly complicated things here. Why don't you use a new class (and metadata field) that stores, for each guideline, an anchor-point and an angle? Those two data already defines a unique straight line. `find_guide_x()` and `find_guide_y()` would be replaced with a single `find_guide()`

regarding the switch to just one find_guide I think yes it makes sense, will look into the best way of doing it. And regarding grouping the coordinates I think that also makes it neater, currently the way it is is that there are about three lists for each ruler (original_cordinate, new coordinates x, new cordinates y) so maybe I can just have all these lists in a struct or something.

rodolforg commented:
I still don't get it, Mohamed :)
Now guides must be described not by a single coordinate, but two coordinates (the reference point / anchor point) and its slope (or angle).
Why you keep lists guide_list_x, guide_list_y, etc? :) And I don't know what "accompanying" means here (I'm not English native speaker, so... sorry XD)

mohamedAdhamc commented:
> I still don't get it, Mohamed :) Now guides must be described not by a single coordinate, but two coordinates (the reference point / anchor point) and its slope (or angle). Why you keep lists guide_list_x, guide_list_y, etc? :) And I don't know what "accompanying" means here (I'm not English native speaker, so... sorry XD)

Well the way I have it is that the rotated ruler is described by two coordinates, the center coordinate and the other coordinate (which I named the accompanying coordinate here). Storing the values of these two coordinates allows us to render them and locate them, etc... We can also use them to calculate various things which I do where it is needed (slope, perpendicular distance, etc...). Now I'm assuming you may be asking as to why I still keep the separated lists of horizontal rulers and vertical rulers now. Well even though they are now both able to be rotated 0-360 degrees, there are still some differences. The biggest and main one being that for vertical rulers the center is always half the height of the canvas, and for horizontal it's half the width, this is to give more freedom with where the ruler can be placed. Am I missing something here :") ?

rodolforg reviewed (changes requested):
This PR can be way simpler and cleaner :)
First, there is a lot of unneeded/temporary/invalid comments.
Second, it is very stateful, and it seems those state variables are not always properly updated.

I propose to create a Guide class, something like this:
```c++
	struct Guide
	{
		synfig::Point point;
		synfig::Angle angle;
	};
```
Because that is all it needs (some other graphic apps, add guide color and label).
Consequently, the GuideList must be changed too.

Then, unify `guide_list_x_` and `guide_list_y_` into a single list: x or y axis can be easily defined by the guide angle/slope.

The `find_guide()` can use the [classic formulas to distance from point to line](https://en.wikipedia.org/wiki/Distance_from_a_point_to_a_line) when iterating the new and single guide list.

You did a good job here, @mohamedAdhamc , but I believe this way it will be more maintainable.

By the way, I can't rotate by GUI, only by the guide editor dialog.

mohamedAdhamc commented:
> 
> I propose to create a Guide class, something like this:

Yes, I also think this will make it more maintainable (although using angles instead of direct points will increase calculations needed but its worth it for maintainability here). Will start the move soon hopefully.

> By the way, I can't rotate by GUI, only by the guide editor dialog.
Maybe it has to do with control not being the same as the mac control ? (I remember it working fine when testing on windows)

rodolforg commented:
> Maybe it has to do with control not being the same as the mac control ? (I remember it working fine when testing on windows)

I use Linux. :)

mohamedAdhamc commented:
> > Maybe it has to do with control not being the same as the mac control ? (I remember it working fine when testing on windows)
> 
> I use Linux. :)

I just moved over to Linux as well :). unfortunately, I haven't had the time much lately to act on the proposed changes. But hopefully when exams/other-work end in about 2 weeks i'll start continuing here ^_^.

mohamedAdhamc commented:
ok, finally I have some time to be able to continue here and implement the guide class suggested. This might take a couple days though as the PR is kinda huge.
I will be working in the following approach
1- go over the whole PR again to remember everything
2- implement guides class and transition everything so it uses it
3- clean up the PR

mohamedAdhamc commented on synfig-studio/src/gui/workarea.h:
```diff
@@ -120,6 +124,8 @@ class WorkArea : public Gtk::Grid, public Duckmatic
 	friend class WorkAreaRenderer;
 	friend class WorkAreaProgress;
 
+	Dialog_Guide guide_dialog;//set up properly
```
done

mohamedAdhamc commented on synfig-studio/src/gui/workarea.h:
```diff
@@ -238,6 +247,8 @@ class WorkArea : public Gtk::Grid, public Duckmatic
 
 	// used in renderer_guides.cpp
 	GuideList::iterator curr_guide;
+	GuideList::iterator curr_guide_accomp; //this need to point to correct accompanying element so we can set its value   //is this used??
+//	AccompGuideList::iterator curr_accomp_guide;
```
done

mohamedAdhamc commented on synfig-studio/src/gui/workarea.cpp:
```diff
@@ -163,12 +163,12 @@ WorkArea::WorkArea(etl::loose_handle<synfigapp::CanvasInterface> canvas_interfac
 	bonesetup_width(0),
-	bonesetup_height(0)
+	bonesetup_height(0),
+	guide_dialog          (*App::main_window,canvas_interface->get_canvas(),this)
```
removed

mohamedAdhamc commented on synfig-studio/src/gui/workarea.cpp:
```diff
@@ -1791,10 +1992,11 @@ WorkArea::on_hruler_event(GdkEvent *event)
 
 			// call the on drawing area event to refresh everything.
-			return on_drawing_area_event(event);
+			return on_drawing_area_event(event); //see what this does
```
done

mohamedAdhamc commented:
@rodolforg finally done :). waiting for your review/testing.

rodolforg commented on synfig-studio/src/gui/duckmatic.h:
```diff
@@ -202,6 +207,10 @@ class Duckmatic
 	bool alternative_mode_;
 	bool lock_animation_mode_;
 
+	bool has_guide_x = false;
+	bool has_guide_y = false;
```
deleted: not used

mohamedAdhamc replied:
it is used in `snap_point_to_grid()` and set from `find_guide()`.

mohamedAdhamc replied:
done

rodolforg commented on synfig-studio/src/gui/workarea.cpp:
```diff
@@ -391,28 +390,21 @@ WorkArea::save_meta_data()
 			if(!data.empty())
 				data+=' ';
-			data+=strprintf("%f",*iter);
+			data+=strprintf("%f", (*iter).point[0]);
+			data+='*';
+			data+=strprintf("%f", (*iter).point[1]);
+			data+='*';
+			data+=strprintf("%f", (*iter).angle.get());
+			data+='*';
+			data+=strprintf("%d", (*iter).isVertical);
```
You can make it into a single line! `"%f*%f` ...

mohamedAdhamc replied:
done

rodolforg commented on synfig-studio/src/gui/workarea.cpp:
```diff
@@ -629,41 +621,34 @@ WorkArea::load_meta_data()
-
-	data=canvas->get_meta_data("guide_y");
-	get_guide_list_y().clear();
+	//for our guide to be stored wed have to store 2 floats x,y + 1 float angle in rad + a true or false
```
`wed`? `we`?

mohamedAdhamc replied:
i think i meant we'd. updated.

rodolforg commented on synfig-studio/src/gui/workarea.cpp:
```diff
@@ -629,41 +621,34 @@ WorkArea::load_meta_data()
-		if(!guide.empty())
-			get_guide_list_y().push_back(stratof(guide));
+		String guide(data.begin(),iter);//this now contains the four things
+		ChangeLocale change_locale(LC_NUMERIC, "C");
```
Move this line to inside `if(!guide.empty()){`  block below.
It temporarily changes locale of the whole program, so it should be placed closest to its usage (and destructed as soon as possible).

mohamedAdhamc replied:
aha ok, done.

rodolforg commented on synfig-studio/src/gui/workarea.cpp:
```diff
@@ -1348,19 +1336,14 @@ WorkArea::on_drawing_area_event(GdkEvent *event)
 						curr_guide = iter;
 						return true;
 					}
+
```
delete?

mohamedAdhamc replied:
done

rodolforg commented on synfig-studio/src/gui/workarea.cpp:
```diff
@@ -1441,14 +1438,17 @@ WorkArea::on_drawing_area_event(GdkEvent *event)
+				drawing_area->queue_draw();
+			}
+
+			if( (iter == get_guide_list().end()) )
+				guide_highlighted = false;
+			else
+				guide_highlighted = true;
```
What about `guide_highlighted = iter != get_guide_list().end()`?

mohamedAdhamc replied:
done

rodolforg commented on synfig-studio/src/gui/workarea.cpp:
```diff
@@ -1536,11 +1555,11 @@ WorkArea::on_drawing_area_event(GdkEvent *event)
-				get_guide_list_x().erase(curr_guide);
+			if(curr_guide->isVertical && !std::isnan(x) && x<0.0 )
+				get_guide_list().erase(curr_guide);
 			else
-			if(!curr_guide_is_x && !std::isnan(y) && y<0.0 )
-				get_guide_list_y().erase(curr_guide);
+			if(!curr_guide->isVertical && !std::isnan(y) && y<0.0 )
+				get_guide_list().erase(curr_guide);
```
Can't rotated guidelines be erased by dragging them back to rulers?

mohamedAdhamc replied:
Yes of course, but vertical guides (guides made from the vertical ruler), are only erased when dragged back to the vertical rulers. Same for horizontal.

rodolforg replied:
Are you saying that if:
- User created a guideline by dragging it from the vertical ruler
- Later it rotates that vertical guideline and make it horizontal
- Drag back to the horizontal ruler it won't be erased?

mohamedAdhamc replied:
done

rodolforg commented on synfig-studio/src/gui/workarea.cpp:
```diff
@@ -1777,15 +1796,16 @@ WorkArea::on_hruler_event(GdkEvent *event)
 	SYNFIG_EXCEPTION_GUARD_BEGIN()
 	switch(event->type) {
 	case GDK_BUTTON_PRESS:
+			from_ruler_event = true;
```
indentantion

mohamedAdhamc replied:
done

rodolforg commented on synfig-studio/src/gui/duckmatic.h:
```diff
@@ -114,6 +114,13 @@ class BezierDrag_Default : public BezierDrag_Base
+	{
+		synfig::Point point;
+		synfig::Angle::rad angle;
+		bool isVertical;
```
`isVertical` is never set. Why do you use it?
Even if it is needed, it should be a method: it's just a matter of check the angle.

mohamedAdhamc replied:
It is only set on ruler creation. It is needed as rulers generated from the vertical ruler have their center of rotations at the middle of the screens height, and the horizontal ruler's have their center of rotations around half the screen width. This is to make the rulers reach more places using just the gui/without the dialog.

rodolforg reviewed (changes requested):
I didn't test the PR yet. I made just a review of part of the code

mohamedAdhamc commented:
updated.

rodolforg commented on synfig-studio/src/gui/duckmatic.cpp:
```diff
@@ -810,38 +810,33 @@ Duckmatic::set_guides_color(const synfig::Color &c)
-Duckmatic::GuideList::iterator
-Duckmatic::find_guide_x(synfig::Point pos, float radius)
+double
+Duckmatic::calculate_distance_from_guide(const Guide guide, const synfig::Point point)
```
Pass argument by reference, not value.

rodolforg commented on synfig-studio/src/gui/duckmatic.cpp:
```diff
@@ -810,38 +810,33 @@ Duckmatic::set_guides_color(const synfig::Color &c)
+Duckmatic::find_guide(synfig::Point pos, float radius)
 {
-	GuideList::iterator iter,best(guide_list_y_.end());
+	has_guide_x = false;
+	has_guide_y = false;
+	GuideList::iterator iter,best(guide_list_.end());
+	second_best_guide_match = guide_list_.end();
```
I still don't understand why `has_guide_x`, `has_guide_y` and `second_best_guide_match` are needed.
Even if they are, they shouldn't be property members of `Duckmatic`.
You are making it stateful:
If somewhere `find_guide()` is called, and another thread call it too, this would screw up these values.

mohamedAdhamc replied:
`has_guide_x` and `has_guide_y` don't need to be class members I removed them. And for`second_best_guide_match` instead of it being a class member I changed it so that if we need to know it we can pass it to the `find_guide` method.

p.s. long story short removed them from being class members

rodolforg commented on synfig-studio/src/gui/duckmatic.cpp:
```diff
@@ -851,35 +846,64 @@ Duckmatic::snap_point_to_grid(const synfig::Point& x)const
-			ret[0]=snap[0],has_guide_x=false;
-		if(std::fabs(snap[1]-ret[1])<=radius && (!has_guide_y || std::fabs(snap[1]-ret[1])<=std::fabs(*guide_y-ret[1])))
-			ret[1]=snap[1],has_guide_y=false;
+		if(std::fabs(snap[0]-ret[0])<=radius && (!has_guide_x || std::fabs(snap[0]-ret[0])<=std::fabs((guide->point)[0]-ret[0])))
```
I believe you should check the distances now the guideline may have any angle:
What is nearest: the vertical grid line, the horizontal grid line or (if guideline snap is enabled) the nearest guideline?

In latter case, then you'll have to compute where is the nearest point in guideline.

mohamedAdhamc replied:
agreed like in the guide snap functionality, I guess I forgot to update this. done.

mohamedAdhamc commented:
ok, now the most recent requested changes should have been addressed, if otherwise please let me know :).

rodolforg commented on synfig-studio/src/gui/workarea.cpp:
```diff
@@ -1405,6 +1381,20 @@ WorkArea::on_drawing_area_event(GdkEvent *event)
 			}
 
+			if(guide_highlighted){
+				Gtk::Menu* waypoint_menu(manage(new Gtk::Menu()));
```
misleading variable name

mohamedAdhamc replied:
Oh yes, definitely xD. Will change to something like 'guide_menu'

mohamedAdhamc replied:
done

rodolforg commented on synfig-studio/src/gui/workarea.cpp:
```diff
@@ -1536,11 +1545,8 @@ WorkArea::on_drawing_area_event(GdkEvent *event)
-			if(!curr_guide_is_x && !std::isnan(y) && y<0.0 )
-				get_guide_list_y().erase(curr_guide);
+			if(!std::isnan(x) && (x<0.0 || y<0.0))
+				get_guide_list().erase(curr_guide);
```
I think if user is rotating guide, it shouldn't be deleted if user release mouse button over a ruler.
However, I tested on Inkscape and it would be deleted there too.

mohamedAdhamc replied:
I think it also makes sense to not delete while rotating but I guess it just come down to user preferences. As you said inkscape deletes in that situation.

P.s. note to self I forgot to add here the check for if y is nan

rodolforg commented on synfig-studio/src/gui/workarearenderer/renderer_guides.cpp:
```diff
@@ -119,44 +105,129 @@ Renderer_Guides::render_vfunc(
+					cr->line_to(
+						x_temp,
+						y_temp
+					);
+					cr->stroke();
+					cr->move_to(x_center,y_center);
+
+					while( (y_temp>0) && (x_temp>0)){
+							x_temp -=1;
+							y_temp +=slope;
+					}
+					cr->line_to(
+						x_temp,
+						y_temp
+					);
+					cr->stroke();
+				}
+			} else {
+				if (iter->isVertical){
+					cr->move_to(
+						x_center,
+						0
+						);
+					cr->line_to(
+						x_center,
+						drawable_h
+					);
+					cr->stroke();
+				} else {
+					cr->move_to(
+						0,
+						y_center
+						);
+					cr->line_to(
+						drawable_w,
+						y_center
+					);
+					cr->stroke();
+				}
+			}
```
You can do it simpler:
Give the guide line parameters (center + angle), you have the equation of the line.
Compute the line points at screen (vertical or horizontal) edges and draw the line that connects them.

mohamedAdhamc replied:
Yes, I thought of a doing something similar but haven't yet. This will better as we will find the two points which make the guide in one step instead of incrementally as it is right now. Will switch over to this implementation in a few hours.

mohamedAdhamc replied:
done.

rodolforg commented on synfig-studio/src/gui/duckmatic.cpp:
```diff
@@ -851,35 +841,69 @@ Duckmatic::snap_point_to_grid(const synfig::Point& x)const
-			ret[0]=snap[0],has_guide_x=false;
-		if(std::fabs(snap[1]-ret[1])<=radius && (!has_guide_y || std::fabs(snap[1]-ret[1])<=std::fabs(*guide_y-ret[1])))
-			ret[1]=snap[1],has_guide_y=false;
+		if ( !has_guide || guide->angle == synfig::Angle::deg(90) || guide->angle == synfig::Angle::deg(0)) {
+			if(std::fabs(snap[0]-ret[0])<=radius && (!has_guide || (guide->angle == synfig::Angle::deg(90) &&
+													 std::fabs(snap[0]-ret[0])<=std::fabs((guide->point)[0]-ret[0]))))
+				ret[0]=snap[0],has_guide=false;
+			if(std::fabs(snap[1]-ret[1])<=radius && (!has_guide || (guide->angle == synfig::Angle::deg(0) &&
+													 std::fabs(snap[1]-ret[1])<=std::fabs((guide->point)[1]-ret[1]))))
+				ret[1]=snap[1],has_guide=false;
+		} else {
+			float distance_from_grid = pow( pow( snap[1] - ret[1], 2.0) + pow(snap[0] - ret[0] , 2.0), 0.5);
+			if ( distance_from_grid <= calculate_distance_from_guide(*guide, ret)){
+				ret = snap;
+				has_guide = false;
+			}
+		}
```
You over complicated it here. :)
The distance from a point to a grid line is just the difference (subtraction) between one of its coordinates and the grid coordinate.
You would have to compare:
- the distance of the point to the nearest horizontal grid line;
- the distance of the point to the nearest vertical grid line; and
- if guideline snap is enabled, the distance of the point to the nearest guide line

If the distance to the nearest 'snapper' is bigger than the acceptance radius, don't snap.
Otherwise, :
- if the 'snapper' is the grid, force the correspondent value of the grid coordinate
- if the 'snapper' is a guideline, compute the nearest point in the guideline

mohamedAdhamc replied:
oh ok, I edited it now but I am still not sure if the new edit is exactly what you mean.

rodolforg commented on synfig-studio/src/gui/duckmatic.cpp:
```diff
@@ -851,35 +841,69 @@ Duckmatic::snap_point_to_grid(const synfig::Point& x)const
-			ret[1]=*guide_y;
+	if(guide_snap && has_guide)
+	{
+		bool possible_intersection_close = second_best_guide_match != nullptr && (second_best_guide_match->angle != guide->angle);
```
Why don't you use the closest guide line to find the point?
Why would the second closest guide line affect the computation?

mohamedAdhamc replied:
The second closest guide line is needed to determine if there is a close intersection of guides to us. The current priority is if there is a close intersection we snap to it. Otherwise we Snap to the closest guide normally. This is mimicking the original behaviour where it would snap to the intersection of the vertical and horizontal rulers. Or should I just ignore snapping to intersections ?

mohamedAdhamc commented:
ok @rodolforg, now the recent requested changes should (maybe) be addressed. 

p.s. hopefully this is getting closer to being ready :)

rodolforg commented on synfig-studio/src/gui/workarea.cpp:
```diff
@@ -1492,10 +1482,29 @@ WorkArea::on_drawing_area_event(GdkEvent *event)
-			else
-				*curr_guide = mouse_pos[1];
+			if (curr_guide->isVertical) {//should then be replaced by checking current guides isVertical
+				if((!rotate_guide && (synfig::Angle::deg(curr_guide->angle).get() == 90))){// case 1: unrotated ruler (vertical)
+					curr_guide->point[0] = mouse_pos[0];
+				} else if (!rotate_guide) { //case 2: rotated ruler being moved
+					float center_x_new= (mouse_pos[0])+((-mouse_pos[1]+curr_guide->point[1])/(tan(curr_guide->angle.get())));
+					curr_guide->point[0] = center_x_new;
+				}
```
Shouldn't both branches give the same result if `curr_guide->angle` is 90°?

mohamedAdhamc replied:
well, currently yes and no. they both behave similarly in a way that if both are 90 degrees they can both be dragged left and right. but for horizontal rulers this makes them lose their center of rotations, they are no longer able to be rotated unless dragged back into correct place. this goes back to the fact that the centers of rotations axis for horizontal rulers and vertical rulers are different (line passing through center of width and center of height).

rodolforg replied:
When a new guide (not the ruler, right?) is created, it will be horizontal or vertical, right? With its initial c.o.r set as the image center, isn't it? I don't even know how to change the c.o.r by mouse (not necessary btw).
So I don't really get the point about this variable.

mohamedAdhamc replied:
I don't know maybe i'm not getting what you mean exactly. But no not always, actually the only case they both have the same c.o.r is at the exact center of the canvas when for example creating by a click not dragging from the ruler. Even then when they are being moved the c.o.r moves differently one moves on x axis other on y (this was to give more flexibility instead of say for example both move only on x).

mohamedAdhamc replied:
Ok, I think I kind of get what your reasoning is (correct me if i'm wrong). You are saying once the rulers are rotated theoretically only the new mouse point is needed (slope is stored) to draw out the guides. This is true for the the guide line equation yes but for the c.o.r it is not we need to know if this ruler is vertical or not to be able to determine how we get the c.o.r. For example if it is we look for intersection with horizontal line passing through half the screen height, for horizontal it is the vertical line passing through half screen width.

So if isVertical isn't stored how would it be determined if this guide is vertical or horizontal to determine the c.o.r ?... Actually I think I kind of get why you think that. Maybe you are thinking horizontal guides become vertical when they are 90 degress so then the isVertical would be outdated info stored ? but this isn't the case a horizontal guide (a guide created from the horizontal ruler) is always a horizontal guide (same c.o.r axis a.lways) even if it becomes vertical (90 degress).

so in a nutshell basically isVertical is used for determining the c.o.r axis vertical or horizontal. A guide created from a vertical ruler is a vertical guide through out its usage and never becomes a horizontal guide even if it look horizontal. 

I think I typed a lot xD, hopefully though now I understand you ?... Also I suggest you can do a little test that will make it realy clear what I mean  if it is possible.
1- create two guides
2- edit their angle to 45 
3- move both to any direction (right for example)
4- notice how the c.o.r (blue plus sign) is moving 

p.s. I will see if I can record the example instead I think that will make it easier.

mohamedAdhamc replied:
An alternative maybe to this method that I just thought of now could be to discard this vertical and horizontal concept altogether and just determine the c.o.r depending on the ruler location. Having it always be the center of the distance between the two furthest guide points (the ones intersecting a canvas axis). 
pros: c.o.r is found in a unified way... always center of the guide length.
cons:
- I think this would  be heavier in calculations as you would be adding the burden of calculating a new c.o.r every time the guide is moved (rotating would keep the ).
-  I just thought of this and I think this just makes the whole idea useless. in certain places the guide c.o.r would be different depending on its angle which would be a huge mess.

mohamedAdhamc replied:
@rodolforg  this is what I meant. For some reason mouse wasn't recorded but basically I was just moving them right then left.

[demoingcor.webm](https://user-images.githubusercontent.com/100296264/221022139-d9366993-0949-4f83-942b-9441469bbb45.webm)

mohamedAdhamc replied:
isVertical removed :)

rodolforg commented on synfig-studio/src/gui/workarea.cpp:
```diff
@@ -1814,28 +1821,28 @@ WorkArea::on_vruler_event(GdkEvent *event)
 	SYNFIG_EXCEPTION_GUARD_BEGIN()
 	switch(event->type) {
 	case GDK_BUTTON_PRESS:
+			from_ruler_event = true;
```
indentation?

mohamedAdhamc replied:
oops, done.

rodolforg commented on synfig-studio/src/gui/workarea.cpp:
```diff
@@ -391,28 +390,15 @@ WorkArea::save_meta_data()
 			if(!data.empty())
 				data+=' ';
-			data+=strprintf("%f",*iter);
+			data+=strprintf(" %f*%f*%f*%d", (*iter).point[0], (*iter).point[1], (*iter).angle.get(), (*iter).isVertical);
```
I thought you said isVertical wasn't necessary anymore. I can't recall it correctly now :/

mohamedAdhamc replied:
No, it is still necessary for the current implementation. The reason being that horizontal and vertical rulers have a fundamental difference. the first having their center of rotation moving up and down at half the screen width, and the latter have their c.o.r moving left and right though half the screen height. This was to allow the rulers to reach more difficult places through GUI.

mohamedAdhamc replied:
isVertical removed :)

mohamedAdhamc commented on synfig-studio/src/gui/workarea.cpp:
```diff
@@ -629,41 +615,34 @@ WorkArea::load_meta_data()
+			else
+				++i;
+		}
+		if(!guide.empty()){
```
I'm not sure i'm catching what you mean here. `guide_components` is declared to be of size 4, so it will always have four elements even if one of them is empty. Maybe you mean to check that if these elements are non empty themselves. But how would this be better ? As long as the guide string isn't empty it's guaranteed to contain the four elements which are always inserted when the metadata is saved/set.

rodolforg replied:
You should at least log it via `synfig::warning()`

rodolforg replied:
Change to `if (guide.size() != 3)`

mohamedAdhamc commented on synfig-studio/src/gui/workarea.cpp:
```diff
@@ -985,6 +964,9 @@ bool
 WorkArea::on_key_press_event(GdkEventKey* event)
 {
 	SYNFIG_EXCEPTION_GUARD_BEGIN()
+	if (event->state == GDK_CONTROL_MASK)
```
unfortunately, I tried all these and the issue still persists. In linux these are all only triggered when ctr+alt are pressed not when ctr alone is pressed like in windows. This is why I was asking if this is possibly a gtk bug, or maybe this is done on purpose due to the nature of the difference of some Linux shortcuts to windows shortcuts.

rodolforg replied:
> unfortunately, I tried all these and the issue still persists. In linux these are all only triggered when ctr+alt are pressed not when ctr alone is pressed like in windows.

When user presses a modifier key only, gtk (on Linux at least - or is it a X11/Wayland stuff?) generates an event with that key itself being pressed, not as a state.
So, check the event `keyval` for `GDK_KEY_Control_L` or `GDK_KEY_Control_R` values.



> why would this flag have to be true while dragging or on hover ?

I meant it should be true only if mouse is hover the guideline or dragging it (and user is holding Ctrl pressed), not every time user hovers or is dragging the guideline. :)

mohamedAdhamc replied:
> 
> When user presses a modifier key only, gtk (on Linux at least - or is it a X11/Wayland stuff?) generates an event with that key itself being pressed, not as a state. So, check the event `keyval` for `GDK_KEY_Control_L` or `GDK_KEY_Control_R` values.

sounds good will do.

> 
> I meant it should be true only if mouse is hover the guideline or dragging it (and user is holding Ctrl pressed), not every time user hovers or is dragging the guideline. :)

Oh ok, now I understand you.

mohamedAdhamc replied:
Done :)

mohamedAdhamc commented on synfig-studio/src/gui/workarea.cpp:
```diff
@@ -1405,6 +1381,20 @@ WorkArea::on_drawing_area_event(GdkEvent *event)
+			if(guide_highlighted){
+				Gtk::Menu* guide_menu(manage(new Gtk::Menu()));
+				guide_menu->signal_hide().connect(sigc::bind(sigc::ptr_fun(&delete_widget), guide_menu));
+				Gtk::MenuItem *item = manage(new Gtk::MenuItem(_("_Edit Guide")));
```
sounds like a good idea. done :).

rodolforg commented:
About the Edit Guide dialog:
- It has that "other point" field. Is it necessary and does it really work?
- If I change the center field and press Apply, it rotates the guide o.O If I press Ok, it changes as expected. My guess is the "Other point" fields.
- The dialog doesn't make clear the units used by Center (and Other) point: pixels, inches or synfig units.

mohamedAdhamc commented:
sounds good, I will hopefully start addressing these issue in about three days. I will also try to take time to make sure everything is clean and finalized so that this could be finally ready for production :).

rodolforg commented:
@mohamedAdhamc ping :)

mohamedAdhamc commented:
> @mohamedAdhamc ping :)

whoops xD, got carried away by other issues, I'll be right back.

mohamedAdhamc commented:
> About the Edit Guide dialog:
> 
>     * It has that "other point" field. Is it necessary and does it really work?
> 
>     * If I change the center field and press Apply, it rotates the guide o.O If I press Ok, it changes as expected. My guess is the "Other point" fields.
> 
>     * The dialog doesn't make clear the units used by Center (and Other) point: pixels, inches or synfig units.

Having coordinates along with the angle in the dialog was supposed to be another way of controlling the rotation, but I agree with you I don't think it is necessary here, typing out a specific angle or rotating by gui should be more than enough (Then maybe this setting guides by points can be added on by a separate PR as an addition after this). So I will remove the part about points altogether from the dialog for now.

rodolforg commented:
renamed PR title because:
![image](https://user-images.githubusercontent.com/1323987/220328537-bd3b1642-3254-4768-8554-9af03834bdde.png)

it rotates the dashed blue *guide lines* (or just guides), not the rulers (the ticked horizontal and vertical bars)

mohamedAdhamc commented:
> renamed PR title

Sounds good ! I made the title when I first started working on this, and I didn't know the difference in terminology at the time (guide vs ruler).

mohamedAdhamc commented:
I commented out the parts related to the setting the guide by points as it is working. It just needs maybe a little modification in behavior or something. Then hopefully after this would be merged I can make a separate small PR for that feature. However, please let me know if It would maybe be better to remove the code instead of commenting.

rodolforg reviewed (changes requested):
Now it works! :)
And now I understand the `isVertical`, but I don't like it XD It can easily disappear.

Last feature-related stuff I suggest here:
1. Double click on a guide pops up the Edit Guide dialog
2. Edit Guide dialog lets user edit the pivot point
3. The pivot point is set wherever user clicks on the guide line (not holding `Ctrl`). Inkscape sets that way, as well as wherever user stops dragging.

The first two items I suggest we left them as new issues for newcomers to Synfig code.

The third one I think it is necessary. What do you think @morevnaproject @ice0?

After that, I'll review the code again (for the last time *fingers crossed*) :)

mohamedAdhamc commented:
> 
> 
> The first two items I suggest we left them as new issues for newcomers to Synfig code.
> 
I like the idea :). Also the second item was implemented here but I removed it, so I could also mention the corresponding commit in the issue as some type of reference for whoever would like to do it.

> After that, I'll review the code again (for the last time _fingers crossed_) :)

let's hope so xD.

rodolforg commented:
We should implement now.
I tested it and it is weird to lose the pivot off screen.

mohamedAdhamc commented:
Alright then, I believe I have time these couple of days so I'll try to do this ASAP.

mohamedAdhamc commented:
@rodolforg 

done :)
I'd like to really thank you for this suggestion. It was rea;;y simple to implement, and it has solved the issue of the pivot going out of bounds. Also simplified the moving guide code. And most importantly removed the need for isVertical.

description of what was done:
1- 
>
>The pivot point is set wherever user clicks on the guide line (not holding Ctrl). Inkscape sets that way, as well as wherever user stops dragging.

2-
removed the `isVertical`

Looking forward to seeing this merged so the users can benefit from it, I believe it now has matured enough. So let's hope after this review it is ready for that ^_^.

rodolforg commented on synfig-studio/src/gui/duckmatic.h:
```diff
@@ -247,6 +251,9 @@ class Duckmatic
 
 	void connect_signals(const Duck::Handle &duck, const synfigapp::ValueDesc& value_desc, CanvasView &canvas_view);
 
+	double calculate_distance_from_guide(const Guide& guide, const synfig::Point& point);
+	double calculate_distance_from_guide(const Guide& guide, const synfig::Point& point)const { return const_cast<Duckmatic*>(this)->calculate_distance_from_guide(guide,point); }
+
```
Looking into the implementation, there is no need to the non-const version.

mohamedAdhamc replied:
oh yes definitely. nice notice :).

rodolforg commented on synfig-studio/src/gui/workarea.cpp:
```diff
@@ -387,28 +386,15 @@ WorkArea::save_meta_data()
+		for (iter = get_guide_list().begin(); iter != get_guide_list().end(); ++iter){
 			if(!data.empty())
 				data+=' ';
-			data+=strprintf("%f",*iter);
+			data+=strprintf(" %f*%f*%f", (*iter).point[0], (*iter).point[1], (*iter).angle.get());
```
1. `(*iter).` can be written as `iter->`
2. if you don't use the iterator itself, but only the value it refers to, you can make a range-based for.
`for (const auto& guide : get_guide_list) {}`

mohamedAdhamc replied:
done

rodolforg commented on synfig-studio/src/gui/workarea.cpp:
```diff
@@ -631,41 +617,34 @@ WorkArea::load_meta_data()
-
-		if(!guide.empty())
-			get_guide_list_y().push_back(stratof(guide));
+		String guide(data.begin(),iter);//this now contains the four items
```
3 items

mohamedAdhamc replied:
done

rodolforg commented on synfig-studio/src/gui/workarea.cpp:
```diff
@@ -1791,28 +1790,28 @@ WorkArea::on_vruler_event(GdkEvent *event)
 		return true;
 	case GDK_BUTTON_RELEASE:
-		if (get_drag_mode() == DRAG_GUIDE && curr_guide_is_x) {
+			from_ruler_event = false;
```
indentation

mohamedAdhamc replied:
done

rodolforg commented on synfig-studio/src/gui/workarea.h:
```diff
@@ -120,6 +124,8 @@ class WorkArea : public Gtk::Grid, public Duckmatic
 	friend class WorkAreaRenderer;
 	friend class WorkAreaProgress;
 
+	Dialog_Guide guide_dialog;
```
Does it need to be public?

mohamedAdhamc replied:
doesn't seem so now at least.

rodolforg commented on synfig-studio/src/gui/dialogs/dialog_guide.h:
```diff
@@ -0,0 +1,99 @@
+
+/* === S T A R T =========================================================== */
+
+#ifndef __SYNFIG_GTKMM_DIALOG_GUIDE_H
+#define __SYNFIG_GTKMM_DIALOG_GUIDE_H
```
Use SYNFIG_GUI_DIALOG_GUIDE_H

mohamedAdhamc replied:
perfecto

rodolforg commented on synfig-studio/src/gui/dialogs/dialog_guide.h:
```diff
@@ -0,0 +1,99 @@
+/* === S Y N F I G ========================================================= */
+/*!	\file dialogs/Dialog_Guide.h
+**	\brief Template Header
```
Please add a simple description on \brief line. "Template Header" is the example of the... template file XD

mohamedAdhamc replied:
oops o_O, done

rodolforg commented on synfig-studio/src/gui/dialogs/dialog_guide.cpp:
```diff
@@ -0,0 +1,261 @@
+/* === S Y N F I G ========================================================= */
+/*!	\file Dialog_Guide.cpp
+**	\brief Template Header
```
Please add a simple description on \brief line. "Template Header" is the example of the... template file XD

mohamedAdhamc replied:
Oh looks like I changed it In the header but forgot the .cpp

mohamedAdhamc replied:
done

rodolforg commented on synfig-studio/src/gui/dialogs/dialog_guide.h:
```diff
@@ -0,0 +1,99 @@
+{
+	typedef std::list<Guide> GuideList;
+
+	etl::handle<synfig::Canvas> canvas;
```
prefer `synfig::Canvas::Handle`

mohamedAdhamc replied:
done

rodolforg commented on synfig-studio/src/gui/dialogs/dialog_guide.h:
```diff
@@ -0,0 +1,99 @@
+	typedef std::list<Guide> GuideList;
+
+	etl::handle<synfig::Canvas> canvas;
+	WorkArea *current_work_area;
+
+	Gtk::SpinButton *angle_widget;
```
Prefer `WorkArea* ` rather `WorkArea *`

mohamedAdhamc replied:
I think this is one of the really old debates in terms of style xD, but ok whichedver works :).

rodolforg commented on synfig-studio/src/gui/dialogs/dialog_guide.h:
```diff
@@ -0,0 +1,99 @@
+public:
+	Dialog_Guide(Gtk::Window& parent, etl::handle<synfig::Canvas> canvas, WorkArea *work_area);
+	~Dialog_Guide();
+	void set_current_guide_and_init(GuideList::iterator current_guide);
```
remove the `_and_init` slice. Setting a guide should change the dialog properly.

rodolforg reviewed (changes requested):
I still have to review the dialog_guide.cpp and duckmatic.cpp aaaand if everything is working XD

I have to sleep now, so I left these points already.

rodolforg commented:
> I'd like to really thank you for this suggestion.

Sorry again for the long journey :/ But it is a really nice feature you are implementing and, as no release is happening for now, better to polish the code and UX soon XD

I hope next dev release contains this feature and eliminates the ETL folder (it is finally coming!).

rodolforg commented on synfig-studio/src/gui/dialogs/dialog_guide.cpp:
```diff
@@ -0,0 +1,261 @@
+	guideGrid->set_row_spacing(6);
+	guideGrid->set_column_spacing(8);
+
+	angle_type_picker.append(_("Degree"));
+	angle_type_picker.append(_("Radian"));
```
use `.append("degree", _("Degree"))`.
Otherwise, the comparison in `set_angle_type()` will depend on translation and fail (as it is now)

rodolforg commented on synfig-studio/src/gui/dialogs/dialog_guide.cpp:
```diff
@@ -0,0 +1,261 @@
+	//Box start
+	Gtk::Box *guide_box = Gtk::manage(new Gtk::Box(Gtk::ORIENTATION_VERTICAL));
+
+	angle_widget=manage(new class Gtk::SpinButton(angle_adjustment,15,2));
```
remove `class` :)

rodolforg commented on synfig-studio/src/gui/dialogs/dialog_guide.cpp:
```diff
@@ -0,0 +1,261 @@
+
+	Gtk::Frame *angleFrame = manage(new Gtk::Frame(_("Rotate By Setting an Angle")));
+	angleFrame->set_shadow_type(Gtk::SHADOW_NONE);
+	((Gtk::Label *) angleFrame->get_label_widget())->set_markup(_("<b>Rotate By Setting an Angle</b>"));
```
prefer `static_cast`

rodolforg commented on synfig-studio/src/gui/dialogs/dialog_guide.cpp:
```diff
@@ -0,0 +1,261 @@
+	point_y_widget->show();
+	point_y_widget->set_hexpand(); */
+
+	Gtk::Frame *angleFrame = manage(new Gtk::Frame(_("Rotate By Setting an Angle")));
```
Prefer `Gtk::Frame* ` rather `Gtk::Frame *`

rodolforg commented on synfig-studio/src/gui/dialogs/dialog_guide.cpp:
```diff
@@ -0,0 +1,261 @@
+
+	Gtk::Button *apply_button(manage(new Gtk::Button(_("_Apply"), true)));
+	apply_button->show();
+	add_action_widget(*apply_button,0);
```
Prefer the enumeration `Gtk::RESPONSE_APPLY` instead of 0.

rodolforg commented on synfig-studio/src/gui/dialogs/dialog_guide.cpp:
```diff
@@ -0,0 +1,261 @@
+
+	Gtk::Button *ok_button(manage(new Gtk::Button(_("_OK"), true)));
+	ok_button->show();
+	add_action_widget(*ok_button,1);
```
Prefer the enumeration `Gtk::RESPONSE_OK` instead of 1.

rodolforg commented on synfig-studio/src/gui/dialogs/dialog_guide.cpp:
```diff
@@ -0,0 +1,261 @@
+//	float rotated_y = center_y + (2.0)*(tan((*curr_guide).angle.get()));
+
+	if(degrees)
+		angle_widget->set_value(synfig::Angle::deg((*curr_guide).angle).get());
```
`curr_guide->angle` instead of `(*curr_guide).angle`

rodolforg commented on synfig-studio/src/gui/dialogs/dialog_guide.cpp:
```diff
@@ -0,0 +1,261 @@
+	if(degrees)
+		angle_widget->set_value(synfig::Angle::deg((*curr_guide).angle).get());
+	else
+		angle_widget->set_value((*curr_guide).angle.get());
```
`curr_guide->angle` instead of `(*curr_guide).angle`

rodolforg commented on synfig-studio/src/gui/dialogs/dialog_guide.cpp:
```diff
@@ -0,0 +1,261 @@
+//	float rotated_x = center_x + 2;
+//	float rotated_y = center_y + (2.0)*(tan((*curr_guide).angle.get()));
+
+	if(degrees)
```
Just for safety, check if `curr_guide` is nullptr

mohamedAdhamc replied:
Actually I believe we dont really have to as `curr_guide` is not a pointer but rather an iterator. What do you think ?

rodolforg replied:
Oh. Although iterators can be invalid, there is no need here indeed :D

rodolforg commented on synfig-studio/src/gui/duckmatic.cpp:
```diff
@@ -815,36 +815,26 @@ Duckmatic::set_guides_color(const synfig::Color &c)
+
 Duckmatic::GuideList::iterator
-Duckmatic::find_guide_y(synfig::Point pos, float radius)
+Duckmatic::find_guide(synfig::Point pos, float radius, Guide* second_best_guide_match)
```
The last argument does not work as intended.
As it is an optional argument and it stores a pointer, you must use `Guide** second_best_guide_match` AND set it in this method something like `*second_best_guide_match = nullptr` (and check if it is not `nullptr` first).

rodolforg replied:
In other words: you are passong the pointer as value, but you pass ot as reference

mohamedAdhamc replied:
> The last argument does not work as intended. As it is an optional argument and it stores a pointer, you must use `Guide** second_best_guide_match` AND set it in this method something like `*second_best_guide_match = nullptr` (and check if it is not `nullptr` first).

This turns out to be a harder task than i anticipated xD. So what I wanted to do here is to pass this pointer and set its value to be null inside the method and then if it is found it becomes not null. this is how i know in the method `snap_point_to_grid` if there is a second best match or not (by checking to see if its not null after it was sent to find_guide). But then if its null I cant set it to point to the iterator without causing a seg error.

Another alternative I tried quickly is to instead switch over to using iterators and setting it to the end of our guide list instead of null but since `snap_point_to_grid` is const its not possible to set the `second_best_guide_match` iterator to be that.

or maybe it is that i'm sleepy rn and missing something o_O.

rodolforg replied:
You can use iterators.

But if you want to understand: https://www.geeksforgeeks.org/passing-reference-to-a-pointer-in-c/

mohamedAdhamc replied:
> You can use iterators.

but then the issue as I mentioned was that I would'nt be able to inialize the iterator to be `guide_list_.end()` in the `snap_point_to_grid` method as it is const.

> But if you want to understand: https://www.geeksforgeeks.org/passing-reference-to-a-pointer-in-c/

I also tried having the argument be a refrence to a pointer but then it cant be null. It seems the only reasonable option here is using a pointer to a pointer, but this is still causing some issues.

so basically the pointer i am sending is null so the argument is now pointing to null and when i try to access the pointer which is being sent I cant because I would be accesing a null pointer. a bit of a closed loop lol.

rodolforg replied:
> I would'nt be able to inialize the iterator to be guide_list_.end()

Try `cend`.


And read the text and my first comment again :)

mohamedAdhamc replied:
> 
> And read the text and my first comment again if you want to use pointers :)

oh yes my bad, I realized what was causing the confusion for me xD. I needed to check if the arg itself was not a nullptr not what it points to.

mohamedAdhamc replied:
Ok, I believe now I found the other issue with snapping to intersection of rotated guides. It still doesn't work "perfectly" though but at least it works like the previous implementation.

What I mean is that if there is no second best guide match there would be no snapping to intersection. second best guide match is present if there is a guide found (distance less than a pre-specified value `radius` in `find_guide`) other than the main current guide.

I will mark this as resolved now but if you feel like there is still an issue feel free to reopen this :)

rodolforg reviewed (changes requested):
It is working great! Thank you!!

However, not the snap to crossing guides :( I proposed the fix above

mohamedAdhamc commented:
> It is working great! Thank you!!
> 
> However, not the snap to crossing guides :( I proposed the fix above

perfect, thanks for taking the time to review all this :), I will try to first address the mentioned changes ASAP (probably by tommrow). Then I will work on fixing snapping to crossing guides. I will then notify you and then after that I believe we should hopefully be done here :).

ice0 commented on synfig-studio/src/gui/dialogs/dialog_guide.cpp:
```diff
@@ -0,0 +1,261 @@
+	point_y_widget->show();
+	point_y_widget->set_hexpand(); */
+
+	Gtk::Frame* angleFrame = manage(new Gtk::Frame(_("Rotate By Setting an Angle")));
```
"Rotate Guide"? I think it will be clearer for translators

mohamedAdhamc replied:
It was originally written this way because there was also an option to rotate through setting the two points that defined the guide. but now the guide is only defined by a pivot and an angle so I guess it could just be "Rotate Guide" then ok.

ice0 replied:
> or would we add setting the guide in the dialog through using two points in the future so then it should be kept as is I'm not sure.

In this case, we'll just rename it back. Perhaps there will be a more understandable name.

P.S. What I mean is that translators only see the phrase to be translated, but they don't see the whole context. So more information might be helpful.

mohamedAdhamc replied:
Ok then I guess for now Rotate Guide would be sufficient.

mohamedAdhamc replied:
done

ice0 commented:
Some thoughts about this PR:
1) What do you think about extracting guide detection code as a separate PR? 
We have issues in MSVC Debug build because of current implementation (current implementation mixing iterators from different lists).
2) UX change for Guide Editor. 
I suggest to rotate guide as soon as the user changes the value in the field.
It will also allow you to change the buttons to `OK`, `Cancel` (which should be more intuitive).

![image](https://user-images.githubusercontent.com/5604544/233896629-638246d9-4630-4f35-9cd2-bfa8d0e178ce.png)

mohamedAdhamc commented:
> 
>     1. What do you think about extracting guide detection code as a separate PR?
>        We have issues in MSVC Debug build because of current implementation (current implementation mixing iterators from different lists).

I think it probably would be too much work as this PR changes many things about the previous implementation of guides. Other than allowing for guide rotation, this PR kind of re-implements guides altogether. So I think having this PR eventually be merged and updating guides at once would probably be better.

>
>     2. UX change for Guide Editor.
>        I suggest to rotate guide as soon as the user changes the value in the field.
>        It will also allow you to change the buttons to `OK`, `Cancel` (which should be more intuitive).
> 
> 
> ![image](https://user-images.githubusercontent.com/5604544/233896629-638246d9-4630-4f35-9cd2-bfa8d0e178ce.png)

Wouldn't this cause history pollution also maybe a weird experience. I can imagine typing 1 3 5 and it rotating to 1 then 13 then 135. Or do you mean something else ?

ice0 commented:
> I think it probably would be too much work as this PR changes many things about the previous implementation of guides. Other than allowing for guide rotation, this PR kind of re-implements guides altogether. So I think having this PR eventually be merged and updating guides at once would probably be better.

Ok.

> Wouldn't this cause history pollution also maybe a weird experience. 

Yes, it makes sense.

> I can imagine typing 1 3 5 and it rotating to 1 then 13 then 135. Or do you mean something else ?

I mean change using `+`/`-` buttons. And save changes in history only when these buttons are released.
Typing should not emit changes, only when the `Enter` key is pressed.

mohamedAdhamc commented:
>
> I mean change using `+`/`-` buttons. And save changes in history only when these buttons are released.
> Typing should not emit changes, only when the `Enter` key is pressed.

Ok I believe this makes more sense now. Maybe one point that I'm still not sure of though is for the +/- buttons I think this would still cause some sort of pollution as users don't always just hold them until they reach a value. I know at least for me i can sometimes click many times. Anyways, what do you think about leaving this for a different PR as well ? I want this PR to be as compact as we could possibly make it so that it eventually gets merged then all additional features as the ones mentioned by Rodolfo could be added ^_^.

ice0 commented:
> Anyways, what do you think about leaving this for a different PR/Issue as well ? I want this PR to be as compact as we could possibly make it so that it eventually gets merged then all additional features as the ones mentioned by Rodolfo could be added ^_^.

Sure.

ice0 commented on synfig-studio/src/gui/duckmatic.cpp:
```diff
@@ -815,36 +815,27 @@ Duckmatic::set_guides_color(const synfig::Color &c)
+
 Duckmatic::GuideList::iterator
-Duckmatic::find_guide_y(synfig::Point pos, float radius)
+Duckmatic::find_guide(synfig::Point pos, float radius, Guide** second_best_guide_match)
```
Can you add a description for this method?
I didn't read the entire discussion, so I didn't understand why the `second_best_guide_match` argument is needed.

mohamedAdhamc replied:
ok, I will add a comment describing it. But yes briefly it is needed in calculations to be able to snap to intersection of two guides.

p.s.
it wasn't needed in previous implementation because the behavior was simple if there is a horizontal guide snap the y coordinate of mouse and if there is a vertical snap the x coordinate of mouse which effectively ended up snapping to the intersection if present.

rodolforg commented on synfig-studio/src/gui/dialogs/dialog_guide.h:
```diff
@@ -0,0 +1,99 @@
+
+/* === S T A R T =========================================================== */
+
+#ifndef __SYNFIG_STUDIO_DIALOG_GUIDE_H
+#define __SYNFIG_STUDIO_DIALOG_GUIDE_H
```
Remove initial `__`

rodolforg commented on synfig-studio/src/gui/dialogs/dialog_guide.cpp:
```diff
@@ -0,0 +1,261 @@
+	canvas(canvas),
+	current_work_area(work_area),
+	angle_adjustment(Gtk::Adjustment::create(0,-2000000000,2000000000,1,1,0)),
+/*	center_x_widget_adjust (Gtk::Adjustment::create(0,-2000000000,2000000000,1,1,0)),
+	center_y_widget_adjust (Gtk::Adjustment::create(0,-2000000000,2000000000,1,1,0)),
+	point_x_widget_adjust (Gtk::Adjustment::create(0,-2000000000,2000000000,1,1,0)),
+	point_y_widget_adjust (Gtk::Adjustment::create(0,-2000000000,2000000000,1,1,0)), */
```
remove all of these commented-out code of dialog

rodolforg commented on synfig-studio/src/gui/dialogs/dialog_guide.cpp:
```diff
@@ -0,0 +1,261 @@
+void
+Dialog_Guide::on_ok_or_apply_pressed(bool ok)
+{
+	if (synfig::Angle::deg(curr_guide->angle).get() != angle_widget->get_value() && degrees) {
+		curr_guide->angle = synfig::Angle::deg(angle_widget->get_value());
+	} else if (curr_guide->angle.get() != angle_widget->get_value() && !degrees) {
+		curr_guide->angle = synfig::Angle::rad(angle_widget->get_value());
+	} /*else
+		set_new_coordinates();*/
```
For readability (and not-needed efficiency), swap the condition order:
if (degrees && ...)

rodolforg commented on synfig-studio/src/gui/duckmatic.cpp:
```diff
@@ -856,35 +847,60 @@ Duckmatic::snap_point_to_grid(const synfig::Point& x)const
+					}
+				}
+		if (!possible_intersection_close && guide->angle == synfig::Angle::deg(90))
+			ret[0]=guide->point[0];
+		else if (!possible_intersection_close && guide->angle == synfig::Angle::deg(0))
+			ret[1]=guide->point[1];
```
surround assignment operator with space ` = `

mohamedAdhamc replied:
the previous code in this method has this style without spaces should I edit that as well ? or maybe refactor the other code in another pr ? because I think having the same method have different styles would probably not be a good thing.

mohamedAdhamc replied:
Alright then, done xD.

rodolforg commented on synfig-studio/src/gui/workarea.cpp:
```diff
@@ -631,41 +617,36 @@ WorkArea::load_meta_data()
+						};
+			get_guide_list().push_back(obj);
+		} else{
+			synfig::warning("Error on trying to restore guide meta data");
```
append: got %d components instead of 3

mohamedAdhamc replied:
I'm replying to the review before this. Checking `guide.size() != 3` wouldn't really make much sense guide is the string which contains the guide meta data which looks something like `guide_center_x_data*guide_center_y_data*guide_angle_data`. I think you confused it with `guide_components` which is the three component vector which we extract the data from guide string into it.

mohamedAdhamc replied:
done, but I am leaving this open in case you mean something else.

rodolforg replied:
> I think you confused it with `guide_components`

:+1:

rodolforg reviewed (changes requested):
Voilà!

rodolforg commented on synfig-studio/src/gui/workarea.cpp:
```diff
@@ -1754,15 +1753,16 @@ WorkArea::on_hruler_event(GdkEvent *event)
-			curr_guide = get_guide_list_y().insert(get_guide_list_y().begin(), 0.0);
-			curr_guide_is_x = false;
+			curr_guide = get_guide_list().insert(get_guide_list().begin(),{synfig::Point(((1.0/2.0)*(drawing_area->get_window()->get_width())*get_pw())+ get_window_tl()[0], 0),
+																			   synfig::Angle::rad(0)});
```
I don't know if we need to change, but FYI there is `synfig::Angle::zero()` :)

mohamedAdhamc replied:
I think it is fine this way. The angle in the `Guide` struct is of type `Angle::rad`. Angle::zero() returns an angle of type `Angle` .

rodolforg reviewed (changes requested):
The last of my part! I swear! XD

mohamedAdhamc commented on synfig-studio/src/gui/duckmatic.cpp:
```diff
@@ -856,35 +847,60 @@ Duckmatic::snap_point_to_grid(const synfig::Point& x)const
+	{
+		bool possible_intersection_close = second_best_guide_match != nullptr && (second_best_guide_match->angle != guide->angle);
+		if (possible_intersection_close){
+					float guides_intersection_x = 0, guides_intersection_y = 0;
+					if (guide->angle != synfig::Angle::deg(90) && second_best_guide_match->angle != synfig::Angle::deg(90)){
+						float slope_best = tan(guide->angle.get()), slope_second_best_guide_match = tan(second_best_guide_match->angle.get());
+						float bestX = guide->point[0], bestY = guide->point[1], lastBestX = second_best_guide_match->point[0], lastBestY = second_best_guide_match->point[1];
+						guides_intersection_x = ( bestY - lastBestY + slope_second_best_guide_match*lastBestX - slope_best*bestX )/(slope_second_best_guide_match - slope_best);
+						guides_intersection_y = slope_best*(guides_intersection_x - bestX) + bestY;
+						possible_intersection_close = (pow(pow(ret[1] - guides_intersection_y, 2.0) + pow(ret[0] - guides_intersection_x, 2.0), 0.5)) <= radius*3;
+					} else if (guide->angle != second_best_guide_match->angle){
+						if (synfig::Angle::deg(guide->angle).get() == 90){
+							guides_intersection_x = guide->point[0];
+							guides_intersection_y = tan(second_best_guide_match->angle.get())*(second_best_guide_match->point[0]) + second_best_guide_match->point[1];
+						} else if (synfig::Angle::deg(second_best_guide_match->angle).get() == 90){
+							guides_intersection_x = second_best_guide_match->point[0];
+							guides_intersection_y = tan(guide->angle.get())*(guide->point[0]) + guide->point[1];
+						}
+						possible_intersection_close = (pow(pow(ret[1] - guides_intersection_y, 2.0) + pow(ret[0] - guides_intersection_x, 2.0), 0.5)) <= radius*3;
+					}
+					if (possible_intersection_close) {
+						ret[0] = guides_intersection_x;
+						ret[1] = guides_intersection_y;
+					}
+				}
```
looks like it wasn't it had too many extra tabs. Done.

rodolforg reviewed (approved):
OMG. Is this real? ;)
@ice0 can you check this right? :D

mohamedAdhamc commented:
> OMG. Is this real? ;) @ice0 can you check this right? :D

I feel like a proud father right now (this PR being my son) xD.

> 
> PS: add good-first-issue for:
> 
>     1. Double click on a guide pops up the Edit Guide dialog
> 
>     2. Edit Guide dialog lets user edit the pivot point

Alright, will do when on this being merged :).

rodolforg commented on synfig-studio/src/gui/duckmatic.h:
```diff
@@ -114,6 +114,12 @@ class BezierDrag_Default : public BezierDrag_Base
 	void bezier_drag(Duckmatic* duckmatic, const synfig::Vector& vector);
 };
 
+struct Guide
+{
+	synfig::Point point;
+	synfig::Angle::rad angle;
+};
+
```
My only doubt in this PR here is if Guide should be defined in duckmatic.h lol

mohamedAdhamc replied:
Hmm I believe this was where the previous guide lists were also defined.

rodolforg commented:
@mohamedAdhamc  looking at the original issue, editing pivot point is required to close it.
So, it's better to implement it here, right? Or in another PR?

mohamedAdhamc commented:
> @mohamedAdhamc looking at the original issue, editing pivot point is required to close it. So, it's better to implement it here, right? Or in another PR?

Oh yes I believe we mentioned here leaving these for other PRs. This one is already kind of huge, so I think it would just be better to leave other additional stuff in new PRs.

mohamedAdhamc commented:
Historical moment lol, I can't believe this is finally merged ^_^. Will make the corresponding new issues shortly.

rodolforg commented:
Merged. Thank you and congrats! 😉

yoSachinkr commented:
@mohamedAdhamc Is there any way to align rotation centers of three or more guides?🤔

mohamedAdhamc commented:
> @mohamedAdhamc Is there any way to align rotation centers of three or more guides?thinking

Well currently whenever you are dragging the guides the new pivot (i.e. center of rotation) is the point you are dragging the guide from. So you can drag them all into one place. And I am supposed to open new issues for some features that didn't make into this PR including being able to set the pivot from the guide dialog (then you could also set as many to have the same center). Or do you mean something else ? :thinking:

yoSachinkr commented:
> So you can drag them all into one place.

I tried doing that but it's not as simple as it sounds. I had to zoom all the way to 12800 times just to make sure their rotation centers are aligned properly. 

> And I am supposed to open new issues for some features that didn't make into this PR (I'll do so in a few hours) including being able to set the pivot from the guide dialog (then you could also set as many to have the same center).

Now this is amazing! 🤩

yoSachinkr commented:
> > > So you can drag them all into one place.
> > 
> > 
> > I tried doing that but it's not as simple as it sounds. I had to zoom all the way to 12800 % just to make sure their rotation centers are aligned properly.
> 
> I can't really remember but I also think having the grid on and snap to grid activated could help.

But right now, guides do not snap to grid. 🤔

mohamedAdhamc commented:
Oh yes this is why I said I cant remember xd. I remember there was snapping between guides and other stuff but I cant remember which and what exactly snapped to what lol. Maybe it was just handles to guides/grids.
